From 1b06675c654727d8aa1545c783482b8e119c64a8 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 278/296] ManagedServices: Catch double unbind of dead service

If a service receives onBindingDied after it has successfully connected
once, unbindService would be called once from the onBindingDied and then
again from registerServiceLocked when the service was rebound.  This
second attempt would typically crash.

The first unbindService call has a try/catch, add one around the second
call as well.
---
 .../core/java/com/android/server/notification/ManagedServices.java  | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/notification/ManagedServices.java b/services/core/java/com/android/server/notification/ManagedServices.java
index 625764cea55..9f6a0d300b0 100644
--- a/services/core/java/com/android/server/notification/ManagedServices.java
+++ b/services/core/java/com/android/server/notification/ManagedServices.java
@@ -844,7 +844,11 @@ abstract public class ManagedServices {
                 Slog.v(TAG, "    disconnecting old " + getCaption() + ": " + info.service);
                 removeServiceLocked(i);
                 if (info.connection != null) {
-                    mContext.unbindService(info.connection);
+                    try {
+                        mContext.unbindService(info.connection);
+                    } catch (IllegalArgumentException e) {
+                        Slog.e(TAG, "failed to unbind " + name, e);
+                    }
                 }
             }
         }
-- 
2.11.0

